### PR TITLE
Arlem is now an actual megafauna and has user-friendly controls

### PR DESCRIPTION
**Pull request by EgorDinamit**
## About The Pull Request

Blast is fixed and now properly shoots at the target.
Damage of blast ability is nerfed.
Charge is now a proper ability and doesn't get fucked up by sandbags/other objects.
Arlem Megafauna has "wastebot" faction as expected.
The megafauna version of Arlem is now placed in the Enclave bunker.

## Why It's Good For The Game

The bunker with one of the most valuable loot has been always a joke in terms of difficulty, because simple mob Arlem was cheeseable as any other mob.
The megafauna version is far from perfect, but it is a proper challenge, unlike what we have now.

## Changelog
:cl:
add: Megafauna Arlem
/:cl:

### CODE REVIEW

**Comment by VEN0M5** [comment]
If you are gonna use megafauna arlem, give him a proper arena with enough space to maneuver.... The current enclave bunker does not have enough space to support a megafauna fight

**Comment by EgorDinamit** [comment]
> If you are gonna use megafauna arlem, give him a proper arena with enough space to maneuver.... The current enclave bunker does not have enough space to support a megafauna fight

It expects you to run outside and fight in the caves, I think.

**Comment by VEN0M5** [comment]
> > If you are gonna use megafauna arlem, give him a proper arena with enough space to maneuver.... The current enclave bunker does not have enough space to support a megafauna fight
> 
> It expects you to run outside and fight in the caves, I think.

No just no, the megafauna was originally meant for Dawson's enclave dungeon. You don't kite a Bubblegum + Colossus hybird across the entire dungeuon that is littered with obstacles... Even worse when you slip on that oil stain at the entrence and die

**Comment by Robloxcumlord** [comment]
> > If you are gonna use megafauna arlem, give him a proper arena with enough space to maneuver.... The current enclave bunker does not have enough space to support a megafauna fight
> 
> It expects you to run outside and fight in the caves, I think.

thats just gonna make it aids to fight if its like bubblegum thats now enough room at all unless ur mining it out

File: _maps/map_files/Pahrump/Pahrump-Underground-1.dmm [hunk 1]
```diff
@@ -15893,7 +15901,15 @@
...
-/mob/living/simple_animal/hostile/enclave/soldier,
+/mob/living/simple_animal/hostile/enclave/soldier{
+	desc = "A former Enclave soldier, mutated by the forced evolutionary virus and grafted to the interior of his suit.";
+	extra_projectiles = 1;
+	faction = list("wastebot");
+	loot = null;
+	name = "Private Harry";
+	speak = list("AUUGHGHHH!");
+	speak_emote = list("shouts")
+	},
```
Comment (Rohesie) [inline _maps/map_files/Pahrump/Pahrump-Underground-1.dmm:15912]: Did you do these var-edits? This should be a custom type instead.

File: code/modules/mob/living/simple_animal/hostile/megafauna/captainarlem.dm [hunk 2]
```diff
@@ -1,40 +1,84 @@
...
+	deathmessage = "chokes on his own blood, falling over as he activates his suit's self destruct!"
 	deathsound = 'sound/effects/gravhit.ogg'
 	del_on_death = FALSE
-	loot = list(/obj/item/keycard/library)
-
+	loot = list(/obj/item/keycard/library, /obj/effect/gibspawner/human, /obj/effect/temp_visual/gib_animation)
+	attack_action_types = list(/datum/action/innate/megafauna_attack/charge,
+							   /datum/action/innate/megafauna_attack/triple_charge,
+							   /datum/action/innate/megafauna_attack/blast,
+							   /datum/action/innate/megafauna_attack/eyebots)
```
Comment (Rohesie) [inline code/modules/mob/living/simple_animal/hostile/megafauna/captainarlem.dm:44]:
```suggestion
	loot = list(
		/obj/item/keycard/library,
		/obj/effect/gibspawner/human,
		/obj/effect/temp_visual/gib_animation,
		)
	attack_action_types = list(
		/datum/action/innate/megafauna_attack/charge,
		/datum/action/innate/megafauna_attack/triple_charge,
		/datum/action/innate/megafauna_attack/blast,
		/datum/action/innate/megafauna_attack/eyebots,
		)
```
Try properly formatting lists and using trailing commas.

File: code/modules/mob/living/simple_animal/hostile/megafauna/captainarlem.dm [hunk 3]
```diff
@@ -1,40 +1,84 @@
...
+	var/charge_cooldown_time = 4 SECONDS
+	var/blast_cooldown_time = 15 SECONDS
+	var/eyebots_cooldown_time = 30 SECONDS
+
+/datum/action/innate/megafauna_attack/charge
+	name = "Charge"
+	icon_icon = 'icons/mob/actions/actions_spells.dmi'
+	button_icon_state = "icebeam0"
+	chosen_message = "<span class='colossus'>You will now dash to your target.</span>"
+	chosen_attack_num = 1
```
Comment (Rohesie) [inline code/modules/mob/living/simple_animal/hostile/megafauna/captainarlem.dm:60]: Could you use defines instead of magic numbers?

File: code/modules/mob/living/simple_animal/hostile/megafauna/captainarlem.dm [hunk 4]
```diff
@@ -90,57 +138,65 @@ mob/living/simple_animal/hostile/megafauna/captainarlem/do_attack_animation(atom
...
+			A.ex_act(EXPLODE_HEAVY)
+		else if(isliving(A))
+			var/mob/living/L = A
+			L.visible_message("<span class='danger'>[src] laughs maniacally as he rends[L]!</span>", "<span class='userdanger'>[src] slams into you, laughing maniacally!</span>")
+			L.apply_damage(50, BRUTE)
+			playsound(get_turf(L), 'sound/effects/meteorimpact.ogg', 100, 1)
+			shake_camera(L, 4, 3)
+			var/throwtarget = get_edge_target_turf(src, get_dir(src, get_step_away(L, src)))
+			L.throw_at(throwtarget, 3)
+	..()
```
Comment (Rohesie) [inline code/modules/mob/living/simple_animal/hostile/megafauna/captainarlem.dm:155]:
```suggestion
	return ..()
```
Preserve parent call return unless you have a reason not to.

File: code/modules/mob/living/simple_animal/hostile/megafauna/captainarlem.dm [hunk 5]
```diff
@@ -90,57 +138,65 @@ mob/living/simple_animal/hostile/megafauna/captainarlem/do_attack_animation(atom
...
 	. = ..()
+
+/mob/living/simple_animal/hostile/megafauna/captainarlem/Bump(atom/A)
 	if(charging)
 		DestroySurroundings()
+		if(isturf(A) || isobj(A) && A.density)
+			A.ex_act(EXPLODE_HEAVY)
+		else if(isliving(A))
+			var/mob/living/L = A
+			L.visible_message("<span class='danger'>[src] laughs maniacally as he rends[L]!</span>", "<span class='userdanger'>[src] slams into you, laughing maniacally!</span>")
```
Comment (Rohesie) [inline code/modules/mob/living/simple_animal/hostile/megafauna/captainarlem.dm:149]: Could you use better variable names than `L`?

File: code/modules/mob/living/simple_animal/hostile/megafauna/captainarlem.dm [hunk 6]
```diff
@@ -152,29 +208,40 @@ mob/living/simple_animal/hostile/megafauna/captainarlem/do_attack_animation(atom
...
 
 /mob/living/simple_animal/hostile/megafauna/captainarlem/death()
 	do_sparks(3, TRUE, src)
-	addtimer(CALLBACK(src, .proc/self_destruct), 4 SECONDS)
-	return ..()
-
-
+	playsound(src, 'sound/weapons/armbomb.ogg', 75, 1, -3)
+	..()
+	sleep(40) // Totally intended to be sleep and not a callback.
```
Comment (Rohesie) [inline code/modules/mob/living/simple_animal/hostile/megafauna/captainarlem.dm:245]: Why this change?

**Comment by Rohesie** [comment]
Also failing the linters.

File: _maps/map_files/Pahrump/Pahrump-Underground-1.dmm (hunk 1 above)
Comment (EgorDinamit) [reply _maps/map_files/Pahrump/Pahrump-Underground-1.dmm:15912]: I just copy-pasted what was there already. Before this chance(as it is now), there is one mob with wrong faction so they end up dying round-start.

File: code/modules/mob/living/simple_animal/hostile/megafauna/captainarlem.dm (hunk 3 above)
Comment (EgorDinamit) [reply code/modules/mob/living/simple_animal/hostile/megafauna/captainarlem.dm:60]: Defines where exactly and for what.

File: code/modules/mob/living/simple_animal/hostile/megafauna/captainarlem.dm (hunk 6 above)
Comment (EgorDinamit) [reply code/modules/mob/living/simple_animal/hostile/megafauna/captainarlem.dm:245]: Some bullshit interaction with an explosion, like destroying loot, etc.

File: code/modules/mob/living/simple_animal/hostile/megafauna/captainarlem.dm (hunk 3 above)
Comment (Rohesie) [reply code/modules/mob/living/simple_animal/hostile/megafauna/captainarlem.dm:66]: Look at `HERALD_TRISHOT`, `TENTACLE_PATCH` and `LEGIONNAIRE_CHARGE` for examples.

File: code/modules/mob/living/simple_animal/hostile/megafauna/captainarlem.dm [hunk 7]
```diff
@@ -152,29 +214,42 @@ mob/living/simple_animal/hostile/megafauna/captainarlem/do_attack_animation(atom
...
 /mob/living/simple_animal/hostile/megafauna/captainarlem/death()
 	do_sparks(3, TRUE, src)
+	playsound(src, 'sound/weapons/armbomb.ogg', 75, 1, -3)
 	addtimer(CALLBACK(src, .proc/self_destruct), 4 SECONDS)
 	return ..()
 
-
+/mob/living/simple_animal/hostile/megafauna/captainarlem/proc/self_destruct()
+	explosion(src,3,5,7,7)
+	qdel(src)
```
Comment (Rohesie) [inline code/modules/mob/living/simple_animal/hostile/megafauna/captainarlem.dm:255]: Put checks here, such as if the mob wasn't qdeleted in the meantime already.
Given it's a timed action, that could have happened.

**Comment by ogrigbe** [comment]
I am afraid of the consequences this may cause if people lead him into the sewers. People getting randomly gibbed after climbing down a ladder or the difficulty that it may cause would be pretty extreme.
